### ISSUE TITLE
Error installing clp on MacOS

### ISSUE THREAD

**Issue by dirkschumacher** [coin-or-tools/homebrew-coinor#61]
I am getting an error during `make test`. Not sure what the problem is and if it is on my side or a bug (probably the former I assume ;)).

```
==> ./configure --prefix=/usr/local/Cellar/clp/1.17.2 --datadir=/usr/local/Cellar/clp/1.17.2/share/clp --includedir=/usr/local/Cellar/clp/1.1
==> make
==> make test
Last 15 lines from /Users/dirkschumacher/Library/Logs/Homebrew/clp/03.make:
71 primal 0 dual -8.88178e-16 low 0 up 1.79769e+308
72 primal 7.46175 dual -1.33227e-15 low 0 up 1.79769e+308
73 primal 0 dual -1.11022e-15 low 0 up 1.79769e+308
74 primal 0.0467292 dual -4.44089e-16 low 0 up 1.79769e+308
75 primal 5.33732 dual -2.88658e-15 low 0 up 1.79769e+308
obj -> -370.86 gradient 0.03
76 primal 0 dual 0.250438 low 0 up 1.79769e+308
77 primal 8.30493 dual 0 low 0 up 1.79769e+308
obj -> -370.113 gradient 0.09
78 primal 16.8962 dual 1.9984e-15 low 0 up 1.79769e+308
obj -> -415.732 gradient -2.7
Computed objective -415.732
/bin/sh: line 1: 73452 Abort trap: 6           ../src/clp -dirSample `echo /usr/local/opt/coin_data_sample/share/coin_data_sample/coin/Data/Sample` -unitTest -dirNetlib `echo /usr/local/opt/coin_data_netlib/share/coin_data_netlib/coin/Data/Netlib` -netlib
make[1]: *** [test] Error 1
make: *** [test] Error 2
```

**Comment by tkralphs** [comment]
All tests pass on OS X build of 1.17.2 on Travis (https://travis-ci.org/coin-or/Clp/builds/522906456) and I have also just installed successfully on my OS X box. I guess we need to dig a bit. I assume you tried to uninstall all dependencies and start from scratch, but if not, try that. Is there any chance there are some old libraries lying around somewhere?

**Comment by dirkschumacher** [comment]
Thanks. Strange, I even tried to compile `gcc` from source. I do get this warning which might cause the problem, not sure. 

```Warning: coin-or-tools/coinor/osi dependency gcc was built with a different C++ standard
library (libstdc++ from clang). This may cause problems at runtime.
```

I will dig further.

**Comment by tkralphs** [comment]
Interesting. I can't get to my OS X box at the moment, but I have consistently seen a similar error, something like `gcc was built with a different C++ standard library...`, which seems to commonly arise in lots of Homebrew installations. Is this what the full error says? It's never caused any problems for me.

**Comment by dirkschumacher** [comment]
Yeah I have seen that warning elsewhere as well. Not sure, but this is the only warning I observe before `make test` of `clp` fails:
```
/bin/sh: line 1: 73452 Abort trap: 6           ../src/clp -dirSample `echo /usr/local/opt/coin_data_sample/share/coin_data_sample/coin/Data/Sample` -unitTest -dirNetlib `echo /usr/local/opt/coin_data_netlib/share/coin_data_netlib/coin/Data/Netlib` -netlib
make[1]: *** [test] Error 1
make: *** [test] Error 2
```

**Comment by dirkschumacher** [comment]
So I unpacked the archive and ran `configure make make test` myself:

```
./configure --prefix=/usr/local/Cellar/clp/1.17.2 --datadir=/usr/local/Cellar/clp/1.17.2/share/clp --includedir=/usr/local/Cellar/clp/1.17.2/include/clp --with-sample-datadir=/usr/local/opt/coin_data_sa
make
make test
```

=>

```
All tests completed successfully
if test -e osiUnitTest ; then \
	  ./osiUnitTest -mpsDir=`echo /usr/local/opt/coin_data_sa` -netlibDir=`echo /usr/local/Cellar/coin_data_netlib/1.2.6/share/coin_data_netlib/coin/Data/Netlib` -testOsiSolverInterface || exit 1 ; \
	fi
Testing OsiRowCut with OsiClpSolverInterface
Coin6001E Unable to open mps input file /usr/local/opt/coin_data_sa/exmip1.mps
Coin0008I  read with -1 errors
*** osirowcutSolverInterface testing issue: read MPS failed: imP->readMps(fn.c_str(), "mps") == 0
Coin6001E Unable to open mps input file /usr/local/opt/coin_data_sa/exmip1.mps
Coin0008I  read with -1 errors
*** osirowcutSolverInterface testing issue: read MPS failed: imP->readMps(fn.c_str(), "mps") == 0
*** osirowcutSolverInterface testing issue: consistent(IntegerModel) failed: cut.consistent(*imP)
Coin6001E Unable to open mps input file /usr/local/opt/coin_data_sa/exmip1.mps
Coin0008I  read with -1 errors
*** osirowcutSolverInterface testing issue: read MPS failed: imP->readMps(fn.c_str(), "mps") == 0
Testing OsiColCut with OsiClpSolverInterface
Coin6001E Unable to open mps input file /usr/local/opt/coin_data_sa/exmip1.mps
Coin0008I  read with -1 errors
Coin6001E Unable to open mps input file /usr/local/opt/coin_data_sa/exmip1.mps
Coin0008I  read with -1 errors
*** osicolcutSolverInterface testing issue: consistent(IntegerModel) failed: cut.consistent(*imP)
*** osicolcutSolverInterface testing issue: consistent(IntegerModel) failed: cut.consistent(*imP)
*** osicolcutSolverInterface testing issue: consistent(IntegerModel) failed: cut.consistent(*imP)
/bin/sh: line 1: 98734 Segmentation fault: 11  ./osiUnitTest -mpsDir=`echo /usr/local/opt/coin_data_sa` -netlibDir=`echo /usr/local/Cellar/coin_data_netlib/1.2.6/share/coin_data_netlib/coin/Data/Netlib` -testOsiSolverInterface
make[1]: *** [test] Error 1
make: *** [test] Error 2
```

**Comment by tkralphs** [comment]
I'm digging in to the unit test a little. The failure on the last few lines is weird (and I would like to know why it's failing), but that itself is not what makes the unit test crash (although the crash and this failure could have a common cause). It looks like memory corruption somewhere.

The test is very simple and I don't see how it could crash (given that we're not having any general problems with Clp or Osi on other platforms). Can you re-configure and re-build with `--enable-debug` to see where it crashes exactly? Even better, if you could run it with `lldb` and see what's happening, that would be great. Or even install valgrind and run it through valgrind.

**Comment by dirkschumacher** [comment]
I get the following error when compiling it with `--enable-debug`

```
Undefined symbols for architecture x86_64:
  "CoinIndexedVector::checkClear()", referenced from:
      ClpFactorization::updateColumnForDebug(CoinIndexedVector*, CoinIndexedVector*, bool) const in libClp.a(ClpFactorization.o)
      ClpSimplexPrimal::pivotResult(int) in libClp.a(ClpSimplexPrimal.o)
      ClpDualRowSteepest::updateWeights(CoinIndexedVector*, CoinIndexedVector*, CoinIndexedVector*, CoinIndexedVector*) in libClp.a(ClpDualRowSteepest.o)
      ClpPESimplex::identifyCompatibleCols(int, int const*, CoinIndexedVector*, CoinIndexedVector*) in libClp.a(ClpPESimplex.o)
      ClpPESimplex::identifyCompatibleRows(CoinIndexedVector*, CoinIndexedVector*) in libClp.a(ClpPESimplex.o)
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [clp] Error 1
make[1]: *** [all] Error 2
make: *** [all-recursive] Error 1
```

```
> clang -v
Apple clang version 11.0.0 (clang-1100.0.33.12)
Target: x86_64-apple-darwin18.7.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
```

**Comment by tkralphs** [comment]
Hmm, I've kind of lost the thread on this. It's a little hard to tell what's happening with that last error except that I guess you must be linking against the wrong version of CoinUtils somehow. I assume you're building outside of homebrew? What libraries are you linking against? Is it possible you have the dependencies from the homebrew core version of clp installed. What does `pkg-config --libs clp` say? Can you attach the `config.log`?

Looking back through the thread and trying to reconstruct, I noticed two things. First, the unit test failure seems to be different when you built from within homebrew and when you built outside of homebrew. Am I seeing that correctly?

Second, I just noticed something in the second failure that I hadn't spotted before, which could actually turn out to be important. There is a line that starts
```
./osiUnitTest -mpsDir=`echo /usr/local/opt/coin_data_sa` ...
```
There is a corruption happening somewhere, since this is not the full path to the sample data directory. This then seems to result in some instance files not being found (which should probably cause the unit test to hard fail, since the rest of it will just be garbage) and then the crash happens shortly after. I can't exactly piece together exactly what might have happened, but can you see what's going on there? Everything still works fine for me.

By the way, there is also a new release of Clp in the meantime, so I could update the recipe, but it might be good to try to untangle what's going on here first.

**Comment by dirkschumacher** [comment]
Ok, I think I finally fixed it. It might have been due to the other `clp` package on `brew`. Not sure. The solution was to delete everything including `coin_data_sample` and `coin_data_netlib`. After that I installed `clp` again using the `coin-or-tools` repository explicitly: `brew install coin-or-tools/coinor/clp`. 
Finally I can continue working cbc and R issues 🙌

**Comment by dirkschumacher** [comment]
@tkralphs thanks a lot for your time looking into that!

**Comment by tkralphs** [comment]
Yes, you unfortunately now need to explicitly `brew install coin-or-tools/coinor/clp`. I myself forget about that when I went back to this just now and and ended up trying to install the clp from homebrew core by accident. This is going to cause endless headaches for me, but I don't know if I can do much about it. I tried to get them to do things in a way that would not cause these issues, but I failed. Anyway, glad you got it fixed. I'll update the recipes to latest versions shortly, so you might want to update once I do that.

FYI, if you haven't been tracking it, Cbc is undergoing some substantial changes, both under the hood and in its build system. I guess there will be a stable release sometime "soon" (interpret liberally) that may (or may not) require some changes to your interface. Are you using the C interface? That is under quite active development due to activity around a new python interface. Stay tuned!
